docs/writing-policies/go/04-validation.md

```diff
@@ -8,331 +8,420 @@ title: ""
-located inside of the `test_data` directory. This directory has already
-been populated by the template repository with an example
-`admissionreview.request` object that matches this tutorial, but you would need
-to craft one with the correct Kubernetes object you are writing the policy for,
-as we covered in ["validating
-policies"](/writing-policies/spec/03-validating-policies.md)
-
-We will start by including the following packages:
+We will replace the contents of the generated file to match the following
+ones:
 
 ```go
+package main
+
 import (
-	"encoding/json"
 	"testing"
 
-	"github.com/deckarep/golang-set"
+	corev1 "github.com/kubewarden/k8s-objects/api/core/v1"
+	metav1 "github.com/kubewarden/k8s-objects/apimachinery/pkg/apis/meta/v1"
+	kubewarden_protocol "github.com/kubewarden/policy-sdk-go/protocol"
 	kubewarden_testing "github.com/kubewarden/policy-sdk-go/testing"
+	"github.com/mailru/easyjson"
 )
-```
 
-The first unit test will ensure that having no user settings leads
-to the request to be accepted:
+func TestValidateLabel(t *testing.T) {
+	// highlight-next-line
+	// NOTE 1
+	cases := []struct {
+		podLabels         map[string]string
+		deniedLabels      []string
+		constrainedLabels map[string]string
+		expectedIsValid   bool
+	}{
+		{
+			// highlight-next-line
+			// Pod has no labels -> should be accepted
+			podLabels:         make(map[string]string),
+			deniedLabels:      []string{"owner"},
+			constrainedLabels: make(map[string]string),
+			expectedIsValid:   true,
+		},
+		{
+			// highlight-next-line
+			// Pod has labels, none is denied -> should be accepted
+			podLabels: map[string]string{
+				"hello": "world",
+			},
+			deniedLabels:      []string{"owner"},
+			constrainedLabels: make(map[string]string),
+			expectedIsValid:   true,
+		},
+		{
+			// highlight-next-line
+			// Pod has labels, one is denied -> should be rejected
+			podLabels: map[string]string{
+				"hello": "world",
+			},
+			deniedLabels:      []string{"hello"},
+			constrainedLabels: make(map[string]string),
+			expectedIsValid:   false,
+		},
+		{
+			// highlight-next-line
+			// Pod has labels, one has constraint that is respected -> should be accepted
+			podLabels: map[string]string{
+				"cc-center": "team-123",
+			},
+			deniedLabels: []string{"hello"},
+			constrainedLabels: map[string]string{
+				"cc-center": `team-\d+`,
+			},
+			expectedIsValid: true,
+		},
+		{
+			// highlight-next-line
+			// Pod has labels, one has constraint that are not respected -> should be rejected
+			podLabels: map[string]string{
+				"cc-center": "team-kubewarden",
+			},
+			deniedLabels: []string{"hello"},
+			constrainedLabels: map[string]string{
+				"cc-center": `team-\d+`,
+			},
+			expectedIsValid: false,
+		},
+		{
+			// highlight-next-line
+			// Settings have a constraint, pod doesn't have this label -> should be rejected
+			podLabels: map[string]string{
```

Comment:
this test case is failing. We reject pods whose constrainedLabels is not satisfied just [if the label is present](https://github.com/kubewarden/docs/blob/c1b22b667903ef269102262803eb1b0588b00646/docs/writing-policies/go/04-validation.md?plain=1#L123-L131), which is not the case in this test.